### DISCUSSION TITLE
Continuous Temperature Prediction Using LSTM with skforecast

### DISCUSSION THREAD

**Discussion by mohsamaz** [JoaquinAmatRodrigo/skforecast#938] [General]
Hello skforecast team,

I am new to skforecast, so I apologize if my question seems a bit basic. I am currently working with a time series dataset (panda Data frame) that has only two columns:

Time in seconds
Temperature 

My goal is to predict the temperature for the next 10 seconds sequentially (i.e., t+1, t+2, …, t+10). After each 10-second interval, my code will have access to new actual data (real t+1) to update the dataset. In essence, I need to continuously predict the next 10 seconds and update the dataset with real-time sensor data as it becomes available.

I plan to use an LSTM model for this task, and I have a few questions:

How should I prepare the data for LSTM?
I suspect that the time column may not be crucial, and I might need to create lagged data from the temperature values instead. Is that correct?

How can I create lagged features from the temperature data that are compatible with LSTM in skforecast?

Does skforecast offer any specific options or classes to handle continuous, recursive predictions along with real-time updates to the dataset?

Thank you very much for your guidance and support!

Best regards,

**Comment by JoaquinAmatRodrigo** [comment]
Hi @mohsamaz,  

Thanks for using Skforecast!  

I believe all your questions are addressed in the documentation:  

1. [Input Data](https://skforecast.org/0.14.0/user_guides/input-data)  
2. [Forecasting with Deep Learning (RNN, LSTM)](https://skforecast.org/0.14.0/user_guides/forecasting-with-deep-learning-rnn-lstm)  
3. [Backtesting](https://skforecast.org/0.14.0/user_guides/backtesting)

If after reading it you have any further doubts, feel free to reach out.

**Comment by mohsamaz** [comment]
@JoaquinAmatRodrigo 
Thank you for your reply and thank you for your wonderful package.
with your help I better understand it.

i tried to do it for me with a random time series:
I dont get the test, prediction plot!
could you please help me why i dont get the plot?

__________________________________________________________________
import random
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
# ==============================================================================
# Function to create a DataFrame with time starting from 0 seconds and 1-second intervals with random temperatures
# ==============================================================================
def create_time_temperature_data(num_entries):
    # Generate time in seconds starting from 0 with 1-second intervals
    time = list(range(num_entries))
    # Generate random temperatures between a range (e.g., 10 to 30 degrees)
    temperature = [random.uniform(10, 30) for _ in range(num_entries)]

    # Create DataFrame
    df = pd.DataFrame({'time': time, 'temperature': temperature})
    return df

# Example usage: create a DataFrame with a specified number of entries
num_entries = 100  # Change this value to get more or fewer entries
df = create_time_temperature_data(num_entries)

plt.plot(df['temperature'])

# Data processing
# ==============================================================================
import os
import pandas as pd
import numpy as np
from skforecast.datasets import fetch_dataset

# Plotting
# ==============================================================================
import matplotlib.pyplot as plt
from skforecast.plot import set_dark_theme
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
import plotly.offline as poff
pio.templates.default = "seaborn"
poff.init_notebook_mode(connected=True)

# Keras
# ==============================================================================
os.environ["KERAS_BACKEND"] = "tensorflow" # 'tensorflow', 'jax´ or 'torch'
import keras
from keras.optimizers import Adam
from keras.losses import MeanSquaredError
from keras.callbacks import EarlyStopping

if keras.__version__ > "3.0":
    if keras.backend.backend() == "tensorflow":
        import tensorflow
    elif keras.backend.backend() == "torch":
        import torch
    else:
        print("Backend not recognized. Please use 'tensorflow' or 'torch'.")

# Time series modeling
# ==============================================================================
import skforecast
from skforecast.deep_learning import ForecasterRnn
from skforecast.deep_learning.utils import create_and_compile_model
from sklearn.preprocessing import MinMaxScaler
from skforecast.model_selection import TimeSeriesFold
from skforecast.model_selection import backtesting_forecaster_multiseries

# Warning configuration
# ==============================================================================
import warnings
warnings.filterwarnings('once')

color = '\033[1m\033[38;5;208m' 
print(f"{color}Version skforecast: {skforecast.__version__}")
print(f"{color}Version Keras: {keras.__version__}")
print(f"{color}Using backend: {keras.backend.backend()}")
print(f"{color}Version pandas: {pd.__version__}")
print(f"{color}Version numpy: {np.__version__}")
if keras.__version__ > "3.0":
    if keras.backend.backend() == "tensorflow":
        print(f"{color}Version tensorflow: {tensorflow.__version__}")
    elif keras.backend.backend() == "torch":
        print(f"{color}Version torch: {torch.__version__}")
    else:
        print(f"{color}Version torch: {jax.__version__}")
        
        
air_quality = df

end_train = 60
end_validation = 80 
air_quality_train = air_quality.loc[:end_train, :].copy()
air_quality_val = air_quality.loc[end_train:end_validation, :].copy()
air_quality_test = air_quality.loc[end_validation:, :].copy()


print(
    f"Dates train      : {air_quality_train.index.min()} --- " 
    f"{air_quality_train.index.max()}  (n={len(air_quality_train)})"
)
print(
    f"Dates validation : {air_quality_val.index.min()} --- " 
    f"{air_quality_val.index.max()}  (n={len(air_quality_val)})"
)
print(
    f"Dates test       : {air_quality_test.index.min()} --- " 
    f"{air_quality_test.index.max()}  (n={len(air_quality_test)})"
)

# Plotting pm2.5
# ==============================================================================
fig, ax = plt.subplots(figsize=(7, 3))
air_quality_train["temperature"].rolling(1).mean().plot(ax=ax, label="train")
air_quality_val["temperature"].rolling(1).mean().plot(ax=ax, label="validation")
air_quality_test["temperature"].rolling(1).mean().plot(ax=ax, label="test")
ax.set_title("temperature")
ax.legend();

# ==============================================================================


# Create model
# ==============================================================================
series = ["temperature"]  # Series used as predictors
levels = ["temperature"]  # Target serie to predict
lags = 5       # Past time steps to be used to predict the target
steps = 5      # Future time steps to be predicted

data = air_quality[series].copy()
data_train = air_quality_train[series].copy()
data_val = air_quality_val[series].copy()
data_test = air_quality_test[series].copy()

model = create_and_compile_model(
    series=data_train,
    levels=levels, 
    lags=lags,
    steps=steps,
    recurrent_layer="LSTM",
    recurrent_units=4,
    dense_units=16,
    optimizer=Adam(learning_rate=0.01), 
    loss=MeanSquaredError()
)
model.summary()

# Forecaster Creation
# ==============================================================================
forecaster = ForecasterRnn(
    regressor=model,
    levels=levels,
    transformer_series=MinMaxScaler(),
    fit_kwargs={
        "epochs": 10,      # Number of epochs to train the model.
        "batch_size": 5,  # Batch size to train the model.
        "callbacks": [
            EarlyStopping(monitor="val_loss", patience=5)
        ],  # Callback to stop training when it is no longer learning.
        "series_val": data_val,  # Validation data for model training.
    },
)    

forecaster



forecaster.fit(data_train)

# Track training and overfitting
# ==============================================================================
fig, ax = plt.subplots(figsize=(5, 2.5))
forecaster.plot_history(ax=ax)



# Backtesting with test data
# ==============================================================================
cv = TimeSeriesFold(
    steps=forecaster.max_step,
    initial_train_size=len(data.loc[:end_validation, :]), # Training + Validation Data
    refit=False,
)
metrics, predictions = backtesting_forecaster_multiseries(
    forecaster=forecaster,
    series=data,
    levels=forecaster.levels,
    cv=cv,
    metric="mean_absolute_error",
    verbose=False, # Set to True for detailed information
)


# Plotting predictions vs real values in the test set
# ==============================================================================

# Plotting predictions vs real values in the test set
# ==============================================================================
fig = go.Figure()
trace1 = go.Scatter(x=data_test.index, y=data_test["temperature"], name="test", mode="lines")
trace2 = go.Scatter(x=predictions.index, y=predictions["temperature"], name="predictions", mode="lines")
fig.add_trace(trace1)
fig.add_trace(trace2)
fig.update_layout(
    title="Prediction vs real values in the test set",
    xaxis_title="Date time",
    yaxis_title="temperature",
    width=750,
    height=350,
    margin=dict(l=20, r=20, t=35, b=20),
    legend=dict(
        orientation="h",
        yanchor="top",
        y=1.05,
        xanchor="left",
        x=0
    )
)
fig.show()

**Comment by JavierEscobarOrtiz** [comment]
Hi,

The `ForecasterRnn` and the `create_and_compile_model` function have been refactored in `skforecast 0.17.0`. 

https://skforecast.org/0.17.0/user_guides/forecasting-with-deep-learning-rnn-lstm

Thank you,

Javi
